client/x/evmstaking/keeper/withdraw.go

```diff
@@ -239,92 +168,148 @@ func (k Keeper) ExpectedRewardWithdrawals(ctx context.Context) ([]RewardWithdraw
 		nextValDelIndex = 0
 
 		// Increase the total swept amount.
-		swept += uint32(len(nextDelegators))
+		swept += uint32(len(nextDelegations))
 	}
 
 	// Update the validator sweep index.
 	if err := k.SetValidatorSweepIndex(ctx, types.NewValidatorSweepIndex(nextValIndex, nextValDelIndex)); err != nil {
-		return nil, err
+		return errors.Wrap(err, "set validator sweep index")
 	}
 
 	log.Debug(
 		ctx, "Finish validator sweep for partial withdrawals",
 		"next_validator_index", nextValIndex,
 		"next_validator_delegation_index", nextValDelIndex,
-		"partial_withdrawals", len(withdrawals),
 	)
 
-	return withdrawals, nil
+	return nil
 }
 
-func (k Keeper) EnqueueEligibleRewardWithdrawal(ctx context.Context, rewardWithdrawals []RewardWithdrawal) error {
-	for i := range rewardWithdrawals {
-		valAddr, err := sdk.ValAddressFromBech32(rewardWithdrawals[i].ValidatorAddress)
-		if err != nil {
-			return errors.Wrap(err, "validator address from bech32")
-		}
+// ProcessEligibleRewardWithdrawal processes the reward withdrawal of delegation.
+func (k Keeper) ProcessEligibleRewardWithdrawal(ctx context.Context, delegation stypes.Delegation, validator stypes.Validator, minRewardWithdrawalAmount uint64) error {
+	// Get validator's address.
+	valBz, err := k.stakingKeeper.ValidatorAddressCodec().StringToBytes(validator.GetOperator())
+	if err != nil {
+		return errors.Wrap(err, "validator address from bech32")
+	}
+	valAddr := sdk.ValAddress(valBz)
+	valAccAddr := sdk.AccAddress(valAddr)
+
+	// Get end current period and calculate rewards.
+	endingPeriod, err := k.distributionKeeper.IncrementValidatorPeriod(ctx, validator)
+	if err != nil {
+		return err
+	}
 
-		valAccAddr := sdk.AccAddress(valAddr).String()
+	delRewards, err := k.distributionKeeper.CalculateDelegationRewards(ctx, validator, delegation, endingPeriod)
+	if err != nil {
+		return err
+	}
 
-		// Withdraw delegation rewards.
-		delAddr := sdk.MustAccAddressFromBech32(rewardWithdrawals[i].DelegatorAddress)
-		delRewards, err := k.distributionKeeper.WithdrawDelegationRewards(ctx, delAddr, valAddr)
+	// if it is self-delegation, add commission
+	if delegation.DelegatorAddress == valAccAddr.String() {
+		// Get validator commissions.
+		valCommission, err := k.distributionKeeper.GetValidatorAccumulatedCommission(ctx, valAddr)
```

Comment:
Instead of getting the validator commission in each iteration of delegation, can we pass it in as an argument to this function? The validator commission is constant per validator during the delegation loop, so it'd be more efficient to retrieve it only once and reduce reads.

Reply:
agreed to @ezreal1997.